File: Packages/CaptureKit/Sources/CaptureKit/CaptureSettings.swift
import Foundation

public enum AfterCaptureBehavior: String, Equatable, CaseIterable {
    case copyOnly, saveOnly, copyAndSave, showOverlay
}

public enum SettingsImageFormat: String, Equatable, CaseIterable {
    case png, jpg
}

public enum OverlayCorner: String, Equatable, CaseIterable {
    case topLeft, topRight, bottomLeft, bottomRight
}

public struct CaptureSettings: Equatable {
    public var afterCapture: AfterCaptureBehavior
    public var format: SettingsImageFormat
    public var overlayCorner: OverlayCorner
    public var overlayAutoDismissSeconds: Int
    public var pinCornerRadius: Int
    public var pinShadow: Bool
    public var historyEnabled: Bool
    public var historyCap: Int
    /// How long a captured screenshot stays in the local cache before its
    /// cached copy + thumbnail are deleted. 0 == Never (keep until the cap evicts it).
    public var historyRetentionSeconds: Int
    public var playSound: Bool

    public static let `default` = CaptureSettings(
        afterCapture: .showOverlay, format: .png,
        overlayCorner: .bottomRight, overlayAutoDismissSeconds: 0)

    public var dictionary: [String: String] {
        ["afterCapture": afterCapture.rawValue,
         "format": format.rawValue,
         "overlayCorner": overlayCorner.rawValue,
         "overlayAutoDismissSeconds": String(overlayAutoDismissSeconds),
         "pinCornerRadius": String(pinCornerRadius),
         "pinShadow": pinShadow ? "true" : "false",
         "historyEnabled": historyEnabled ? "true" : "false",
         "historyCap": String(historyCap),
         "historyRetentionSeconds": String(historyRetentionSeconds),
         "playSound": playSound ? "1" : "0"]
    }

    public init(afterCapture: AfterCaptureBehavior, format: SettingsImageFormat,
                overlayCorner: OverlayCorner, overlayAutoDismissSeconds: Int,
                pinCornerRadius: Int = 8, pinShadow: Bool = true,
                historyEnabled: Bool = true, historyCap: Int = 50,
                historyRetentionSeconds: Int = 1800,
                playSound: Bool = true) {
        self.afterCapture = afterCapture
        self.format = format
        self.overlayCorner = overlayCorner
        self.overlayAutoDismissSeconds = overlayAutoDismissSeconds
        self.pinCornerRadius = pinCornerRadius
        self.pinShadow = pinShadow
        self.historyEnabled = historyEnabled
        self.historyCap = historyCap
        self.historyRetentionSeconds = historyRetentionSeconds
        self.playSound = playSound
    }

    public init(dictionary: [String: String]) {
        let d = CaptureSettings.default
        self.afterCapture = AfterCaptureBehavior(rawValue: dictionary["afterCapture"] ?? "") ?? d.afterCapture
        self.format = SettingsImageFormat(rawValue: dictionary["format"] ?? "") ?? d.format
        self.overlayCorner = OverlayCorner(rawValue: dictionary["overlayCorner"] ?? "") ?? d.overlayCorner
        let rawAutoDismiss = Int(dictionary["overlayAutoDismissSeconds"] ?? "") ?? d.overlayAutoDismissSeconds
        // Snap to the slider's stop table so a legacy persisted value (e.g. an older
        // build's 6s default) resolves to a stop the Settings slider can show.
        self.overlayAutoDismissSeconds = OverlayDismissScale.snap(rawAutoDismiss)
        self.pinCornerRadius = Int(dictionary["pinCornerRadius"] ?? "") ?? d.pinCornerRadius
        self.pinShadow = dictionary["pinShadow"].map { $0 == "true" } ?? d.pinShadow
        self.historyEnabled = dictionary["historyEnabled"].map { $0 == "true" } ?? d.historyEnabled
        let rawHistoryCap = Int(dictionary["historyCap"] ?? "") ?? d.historyCap
        // Snap to the allowed option set so a legacy persisted value (e.g. an
        // older build's default of 200) still resolves to a value the
        // Settings UI's 10/50/100 control can show as selected.
        self.historyCap = [10, 50, 100].min(by: { abs($0 - rawHistoryCap) < abs($1 - rawHistoryCap) }) ?? 50
        let rawRetention = Int(dictionary["historyRetentionSeconds"] ?? "") ?? d.historyRetentionSeconds
        // Snap to the slider's stop table, same as overlayAutoDismissSeconds above.
        self.historyRetentionSeconds = HistoryRetentionScale.snap(rawRetention)
        self.playSound = (dictionary["playSound"] ?? "1") != "0"
    }
}

File: windows/src/BetterScreenshot.Capture/CaptureSettings.cs
using System.Globalization;

namespace BetterScreenshot.Capture;

public enum AfterCaptureBehavior { CopyOnly, SaveOnly, CopyAndSave, ShowOverlay }
public enum SettingsImageFormat { Png, Jpg }
public enum SettingsOverlayCorner { TopLeft, TopRight, BottomLeft, BottomRight }

/// <summary>
/// Capture behavior settings, persisted as a flat string dictionary (1:1 with the macOS app's persisted keys).
/// Defaults: show the quick-access overlay, PNG, bottom-right corner, 30s auto-dismiss, pin radius 8 + shadow,
/// history enabled with a 50-item cap.
/// </summary>
public sealed record CaptureSettings
{
    public AfterCaptureBehavior AfterCapture { get; init; } = AfterCaptureBehavior.ShowOverlay;
    public SettingsImageFormat Format { get; init; } = SettingsImageFormat.Png;
    public SettingsOverlayCorner OverlayCorner { get; init; } = SettingsOverlayCorner.BottomRight;
    public int OverlayAutoDismissSeconds { get; init; } = 30;
    public int PinCornerRadius { get; init; } = 8;
    public bool PinShadow { get; init; } = true;
    public bool HistoryEnabled { get; init; } = true;
    public int HistoryCap { get; init; } = 50;

    public static CaptureSettings Default => new();

    public Dictionary<string, string> ToDictionary() => new()
    {
        ["afterCapture"] = AfterCapture switch
        {
            AfterCaptureBehavior.CopyOnly => "copyOnly",
            AfterCaptureBehavior.SaveOnly => "saveOnly",
            AfterCaptureBehavior.CopyAndSave => "copyAndSave",
            _ => "showOverlay",
        },
        ["format"] = Format == SettingsImageFormat.Jpg ? "jpg" : "png",
        ["overlayCorner"] = OverlayCorner switch
        {
            SettingsOverlayCorner.TopLeft => "topLeft",
            SettingsOverlayCorner.TopRight => "topRight",
            SettingsOverlayCorner.BottomLeft => "bottomLeft",
            _ => "bottomRight",
        },
        ["overlayAutoDismissSeconds"] = OverlayAutoDismissSeconds.ToString(CultureInfo.InvariantCulture),
        ["pinCornerRadius"] = PinCornerRadius.ToString(CultureInfo.InvariantCulture),
        ["pinShadow"] = PinShadow ? "true" : "false",
        ["historyEnabled"] = HistoryEnabled ? "true" : "false",
        ["historyCap"] = HistoryCap.ToString(CultureInfo.InvariantCulture),
    };

    public static CaptureSettings FromDictionary(IReadOnlyDictionary<string, string> d)
    {
        var def = Default;
        return new CaptureSettings
        {
            AfterCapture = d.TryGetValue("afterCapture", out var ac)
                ? ac switch
                {
                    "copyOnly" => AfterCaptureBehavior.CopyOnly,
                    "saveOnly" => AfterCaptureBehavior.SaveOnly,
                    "copyAndSave" => AfterCaptureBehavior.CopyAndSave,
                    "showOverlay" => AfterCaptureBehavior.ShowOverlay,
                    _ => def.AfterCapture,
                }
                : def.AfterCapture,
            Format = d.TryGetValue("format", out var f) ? (f == "jpg" ? SettingsImageFormat.Jpg : SettingsImageFormat.Png) : def.Format,
            OverlayCorner = d.TryGetValue("overlayCorner", out var oc)
                ? oc switch
                {
                    "topLeft" => SettingsOverlayCorner.TopLeft,
                    "topRight" => SettingsOverlayCorner.TopRight,
                    "bottomLeft" => SettingsOverlayCorner.BottomLeft,
                    "bottomRight" => SettingsOverlayCorner.BottomRight,
                    _ => def.OverlayCorner,
                }
                : def.OverlayCorner,
            // Snap to the slider's stop table so a legacy persisted value (e.g. an older build's 6s
            // default) resolves to a stop the Settings slider can show.
            OverlayAutoDismissSeconds = OverlayDismissScale.Snap(
                ParseInt(d, "overlayAutoDismissSeconds", def.OverlayAutoDismissSeconds)),
            PinCornerRadius = ParseInt(d, "pinCornerRadius", def.PinCornerRadius),
            PinShadow = ParseBool(d, "pinShadow", def.PinShadow),
            HistoryEnabled = ParseBool(d, "historyEnabled", def.HistoryEnabled),
            HistoryCap = ParseInt(d, "historyCap", def.HistoryCap),
        };
    }

    private static int ParseInt(IReadOnlyDictionary<string, string> d, string key, int fallback) =>
        d.TryGetValue(key, out var v) && int.TryParse(v, NumberStyles.Integer, CultureInfo.InvariantCulture, out var n) ? n : fallback;

    private static bool ParseBool(IReadOnlyDictionary<string, string> d, string key, bool fallback) =>
        d.TryGetValue(key, out var v) ? v == "true" : fallback;
}
